### DISCUSSION TITLE
Do I need a full rerdraw each time when using skcontext?

### DISCUSSION THREAD

**Discussion by Jon-Murray** [AvaloniaUI/Avalonia#9150] [Q&A]
Heya, I've just started using avalonia preview 2 and have been playing around with the Skia custom drawing implementations. All of the reference material and the examples I've seen online seem to indicate that you need to do a full redraw of the image each time the control is invalidated. I have the situation where I have a very large SKBitmap, which small regions are updated. Using the generic code i've seen around:

``` 
public void Render(IDrawingContextImpl context)
            {
                var leaseFeature = context.GetFeature<ISkiaSharpApiLeaseFeature>();

                if (leaseFeature is null)
                {
                    throw new Exception("SkiaSharp is not supported.");
                }

                using var lease = leaseFeature.Lease();

                lease.SkCanvas.draw... 
```

will (I believe) always do a full redraw of the bitmap, even if only 1px has changed. I've found a few references online as to how to tile images using Skia, do region updates/overwrites in Skia, and how to do writeable bitmaps in Avalonia (all techniques would be fine). However, I've been unable to find any examples to try and do what I want (which is update only a specific region of an SKBitmap, without performing a full redraw). I've toyed with the idea of splitting the skbitmap into 10x10 regions, but then I'd need 100 usercontrols/canvases in Avalonia in order to update the specific regions, I guess?

So far, I have a combination of creating my own SKSurface, which I pass into my ICustomDrawOperation each time the outer render is called... this successfully works in that I am able to continually update regions of the larger bmp without having to parse/create/do multiple updates of it each time... but then I still have to have the final ```lease.SkCanvas.draw``` to get it to render out properly to the page. 

Are there any examples anywhere I can look at about the best way to perform partial updates in Avalonia/Skia please? It doesn't necessarily need to be within the context of a UserControl/ICustomDrawOperation, I'm just really coming up short after several days of searching/trying things out on my own. Thanks!

**Comment by maxkatz6** [comment]
The problem that controls partially behind your custom draw operation might be redrawn. And as a result, custom draw operation should be fully redrawn to stay on top.
But you need to know, that it won't be redrawn, if invalidated dirty rect wasn't touching area of the app with your custom draw operation.
You can see dirty rects in the app by enabling `window.PlatformImpl.Renderer.ShowDirtyRects`.

Currently our API doesn't provide dirty rects to the custom draw operation.
Alternatively, you can renderer changes into an intermediate skia bitmap, and render only that in the custom draw operation. Which, I guess, you are already doing. Other than that you might want to keep size of the bitmap as close as possible to how it will be rendered.
And additionally - [increase skia max allowed GPU memory](https://github.com/AvaloniaUI/Avalonia/issues/3043#issuecomment-624774905), as it might get slow otherwise.
